llvm/lib/SYCL/KernelProperties.cpp

```diff
@@ -0,0 +1,66 @@
+//===- KernelProperties.cpp -----------------------------------------------------===//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+//===----------------------------------------------------------------------===//
+//
+// Contains utility functions that read sycl kernel properties
+// ===---------------------------------------------------------------------===//
+
+#include "llvm/Analysis/ValueTracking.h"
+#include "llvm/IR/GlobalVariable.h"
+#include "llvm/IR/InstIterator.h"
+#include "llvm/SYCL/KernelProperties.h"
+
+using namespace llvm;
+namespace {
+  static StringRef KindOf(const char *Str) {
+    return StringRef(Str, strlen(Str) + 1);
+  }
+}
+
+namespace llvm {
+KernelProperties::KernelProperties(Function &F) {
+  for (Instruction &I : instructions(F)) {
+    auto *CB = dyn_cast<CallBase>(&I);
+    if (!CB || CB->getIntrinsicID() != Intrinsic::var_annotation)
+      continue;
+    auto *Alloca =
+        dyn_cast_or_null<AllocaInst>(getUnderlyingObject(CB->getOperand(0)));
+    auto *Str = cast<ConstantDataArray>(
+        cast<GlobalVariable>(getUnderlyingObject(CB->getOperand(1)))
+            ->getOperand(0));
+    if (!Alloca)
+      continue;
+    if (Str->getRawDataValues() != KindOf("xilinx_ddr_bank"))
```

Comment:
Probably this code needs some comments in the neighborhoods and also some feedback in the case the decoration `xilinx_ddr_bank` is misused (warning message...).